iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/ArgDefinedOutsideOfScopeCheck.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * SonarQube IaC Plugin
+ * Copyright (C) 2021-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.docker.checks;
+
+import java.util.Objects;
+import java.util.stream.Stream;
+import org.sonar.check.Rule;
+import org.sonar.iac.common.api.checks.CheckContext;
+import org.sonar.iac.common.api.checks.IacCheck;
+import org.sonar.iac.common.api.checks.InitContext;
+import org.sonar.iac.docker.symbols.Scope;
+import org.sonar.iac.docker.symbols.Usage;
+import org.sonar.iac.docker.tree.api.DockerTree;
+import org.sonar.iac.docker.tree.api.Variable;
+
+@Rule(key = "S6579")
+public class ArgDefinedOutsideOfScopeCheck implements IacCheck {
+  private static final String MESSAGE = "Include the ARG instruction in the build stage where it is used.";
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(Variable.class, ArgDefinedOutsideOfScopeCheck::checkVariableIsDefinedLocally);
+  }
+
+  private static void checkVariableIsDefinedLocally(CheckContext ctx, Variable variable) {
+    if (variable.symbol() == null || Stream.iterate(variable, Objects::nonNull, DockerTree::parent).anyMatch(t -> t.is(DockerTree.Kind.FROM))) {
+      // Variable is either not a Dockerfile variable or is path of FROM instruction where it doesn't need to be redeclared
+      return;
+    }
+
+    boolean isDefinedInImageScope = false;
+    boolean isDefinedInGlobalScope = false;
+    for (Usage usage : variable.symbol().usages()) {
+      if (usage.kind().equals(Usage.Kind.ASSIGNMENT)) {
+        isDefinedInImageScope = usage.scope().kind().equals(Scope.Kind.IMAGE);
+        isDefinedInGlobalScope = usage.scope().kind().equals(Scope.Kind.GLOBAL);
+      }
+    }
+    boolean isDefinedOnlyInGlobalScope = isDefinedInGlobalScope && !isDefinedInImageScope;
+    if (isDefinedOnlyInGlobalScope) {
```

Comment:
This code is a bit strange.

You are overriding `isDefinedInImageScope` and `isDefinedInGlobalScope` at every loop iteration, which also means they can never have the same value, so the boolean `isDefinedOnlyInGlobalScope` is useless.

From my understanding, you are trying to check if a variable is only defined with a `Global` scope.
You don't really need to do this, but only to check if the last assignment is `Global`, as they are kept in the line order.
```suggestion
    
    List<Usage> usages = variable.symbol().usages();
    boolean isGlobalScope = false;
    for (int i = usages.size() - 1; i >= 0; i--) {
      Usage usage = usages.get(i);
      if (usage.kind().equals(Usage.Kind.ASSIGNMENT)) {
        isGlobalScope = usage.scope().kind().equals(Scope.Kind.GLOBAL);
        break; // we stop at the first assignment we find
      }
    }

    if (isGlobalScope) {
```
(Feel free to improve the suggested code, I couldn't find a better way to write it).

Reply:
Thanks for catching this, my code is definitely incorrect here :sweat_smile: 
However, I think that relying on the order in which usages are present couples the check logic too tightly with symbol table implementation. I would rather fix my code or add an accessor `Symbol#getDeclarationScope`